bika/lims/idserver.py

```diff
@@ -43,154 +41,125 @@ def idserver_generate_id(context, prefix, batch_size = None):
-
-    # Analysis Request IDs
-    if context.portal_type == "AnalysisRequest":
-        sample = context.getSample()
-        s_prefix = fn_normalize(sample.getSampleType().getPrefix())
-        sample_padding = context.bika_setup.getSampleIDPadding()
-        ar_padding = context.bika_setup.getARIDPadding()
-        sample_id = sample.getId()
-        sample_number = sample_id.split(s_prefix)[1]
-        ar_number = sample.getLastARNumber()
-        ar_number = ar_number and ar_number + 1 or 1
-        return fn_normalize(
-            ("%s%s" + separator + "R%s") % (s_prefix,
-                          str(sample_number).zfill(sample_padding),
-                          str(ar_number).zfill(ar_padding))
-        )
-
-    # Sample Partition IDs
-    if context.portal_type == "SamplePartition":
-        # We do not use prefixes.  There are actually codes that require the 'P'.
-        # matches = [p for p in prefixes if p['portal_type'] == 'SamplePartition']
-        # prefix = matches and matches[0]['prefix'] or 'samplepartition'
-        # padding = int(matches and matches[0]['padding'] or '0')
-
-        # at this time the part exists, so +1 would be 1 too many
-        partnr = str(len(context.aq_parent.objectValues('SamplePartition')))
-        # parent id is normalized already
-        return ("%s" + separator + "P%s") % (context.aq_parent.id, partnr)
-
-    if context.bika_setup.getExternalIDServer():
-
-        # if using external server
-
-        for d in prefixes:
-            # Sample ID comes from SampleType
-            if context.portal_type == "Sample":
-                prefix = context.getSampleType().getPrefix()
-                padding = context.bika_setup.getSampleIDPadding()
-                new_id = str(idserver_generate_id(context, "%s%s-" % (prefix, year)))
-                if padding:
-                    new_id = new_id.zfill(int(padding))
-                return ('%s%s' + separator + '%s') % (prefix, year, new_id)
-            elif d['portal_type'] == context.portal_type:
-                prefix = d['prefix']
-                padding = d['padding']
-                new_id = str(idserver_generate_id(context, "%s%s-" % (prefix, year)))
-                if padding:
-                    new_id = new_id.zfill(int(padding))
-                return ('%s%s' + separator + '%s') % (prefix, year, new_id)
-        # no prefix; use portal_type
-        # year is not inserted here
-        # portal_type is be normalized to lowercase
-        npt = id_normalize(context.portal_type)
-        new_id = str(idserver_generate_id(context, npt + "-"))
-        return ('%s' + separator + '%s') % (npt, new_id)
+    if portal_type == '':
+        portal_type = context.portal_type
 
+    def getLastCounter(context, config):
+        if config.get('counter_type', '') == 'backreference':
+            relationship = config['counter_reference']
+            backrefs = get_backreferences(context, relationship)
+            return len(backrefs) - 1
```

Comment:
IDServer doc test, while creating Analysis Requests, fail:
```
Differences (ndiff with -expected +actual):
    - <...water17-0001-R1>
    + <AnalysisRequest at /plone/clients/client-1/water17-0001-R0>
```

`backrefs` length is always 0 and the function ends up returning -1. I've been walking through `uidreferencefield`, checking `get_backreferences`, `_set_backreferences`, etc. before and after the AR's `processForm` in `create_analysisrequest` and I always get valid backreferences from Sample. The weird thing is that I've tested IDServer running the instance and all works perfect.... so I started to think about `transaction.savepoint()`, etc... but w/o success. The UIDReferenceField doc test works too, so I'm a bit lost here........ I'll keep digging.

Reply:
Calling `get_backreferences(obj.getSample())` after this:

https://github.com/rockfruit/bika.lims/blob/1f0fc0fefa0e3719e46e2510708f8400cf851190/bika/lims/idserver.py#L176-L177

Investigating now what the hell is going on with this savepoint...

Reply:
From @ramonski in gitter
I think the issue regarding the backreferences in IDServer is located here: 
https://github.com/senaite/bika.lims/blob/senaite-integration/bika/lims/browser/fields/uidreferencefield.py#L144

This is a simple list type and Zope does not recognize for these types when the data has changed.

https://docs.plone.org/develop/plone/persistency/persistent.html#lists-and-dictionaries

Fixed here:
https://github.com/senaite/bika.lims/commit/93cd343a2bab4b76a32b210ef8ee28bf30d773a1

AFAIK mutable objects are read/written in one transaction completely, becuase the ZODB does not recognize that these have changed

Therfore the `_p_changed = True` dance